# Fraud Detection using Machine Learning

---

> **Notebook:** `01_Business_Understanding.ipynb`  
> **Dataset:** Kaggle — Credit Card Fraud Detection  
> **Last Updated:** July 2026

---

In [10]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

In [11]:
# Load dataset — raw string to avoid backslash escape issues
DATA_PATH = Path(r"D:\Fraud Detection\data\raw\creditcard.csv")

df = pd.read_csv(DATA_PATH)
print(f"Dataset loaded: {df.shape[0]:,} rows x {df.shape[1]} columns")

Dataset loaded: 284,807 rows x 31 columns


## 1. Project Overview


In [12]:
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0000,-1.3598,-0.0728,2.5363,1.3782,-0.3383,0.4624,0.2396,0.0987,0.3638,0.0908,-0.5516,-0.6178,-0.9914,-0.3112,1.4682,-0.4704,0.2080,0.0258,0.4040,0.2514,-0.0183,0.2778,-0.1105,0.0669,0.1285,-0.1891,0.1336,-0.0211,149.6200,0
1,0.0000,1.1919,0.2662,0.1665,0.4482,0.0600,-0.0824,-0.0788,0.0851,-0.2554,-0.1670,1.6127,1.0652,0.4891,-0.1438,0.6356,0.4639,-0.1148,-0.1834,-0.1458,-0.0691,-0.2258,-0.6387,0.1013,-0.3398,0.1672,0.1259,-0.0090,0.0147,2.6900,0
2,1.0000,-1.3584,-1.3402,1.7732,0.3798,-0.5032,1.8005,0.7915,0.2477,-1.5147,0.2076,0.6245,0.0661,0.7173,-0.1659,2.3459,-2.8901,1.1100,-0.1214,-2.2619,0.5250,0.2480,0.7717,0.9094,-0.6893,-0.3276,-0.1391,-0.0554,-0.0598,378.6600,0
3,1.0000,-0.9663,-0.1852,1.7930,-0.8633,-0.0103,1.2472,0.2376,0.3774,-1.3870,-0.0550,-0.2265,0.1782,0.5078,-0.2879,-0.6314,-1.0596,-0.6841,1.9658,-1.2326,-0.2080,-0.1083,0.0053,-0.1903,-1.1756,0.6474,-0.2219,0.0627,0.0615,123.5000,0
4,2.0000,-1.1582,0.8777,1.5487,0.4030,-0.4072,0.0959,0.5929,-0.2705,0.8177,0.7531,-0.8228,0.5382,1.3459,-1.1197,0.1751,-0.4514,-0.2370,-0.0382,0.8035,0.4085,-0.0094,0.7983,-0.1375,0.1413,-0.2060,0.5023,0.2194,0.2152,69.9900,0


In [15]:
df.sample(10)

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
252873,156005.0000,-0.7276,0.8172,-0.2234,-1.7001,0.5825,-0.4956,0.3465,0.2977,-1.7803,-0.7858,-1.3131,-1.5596,-0.4048,-0.9674,-0.0503,1.1554,1.4529,-0.7463,1.4330,0.1293,-0.0458,-0.4016,-0.4342,-0.0166,0.8410,-0.0621,-0.1178,-0.0079,12.0000,0
30428,35918.0000,1.3766,-0.7357,1.1591,-0.5023,-1.6219,-0.5053,-1.1728,0.0189,-0.0737,0.5167,-0.5858,-0.5417,0.0944,-0.6881,0.3731,1.1889,0.3199,-1.0109,0.4635,0.0444,0.2998,0.9571,-0.0924,0.4460,0.4558,-0.0453,0.0509,0.0236,0.0000,0
169916,119928.0000,1.8906,-1.3778,0.6461,0.8484,-1.4099,1.7934,-2.0064,0.6464,1.1007,0.7660,-0.7688,1.1074,0.3513,-1.2736,-2.3459,-1.4354,-0.1764,2.1015,-0.5190,-0.5754,-0.2059,0.3498,0.1537,0.2073,-0.2100,-0.4957,0.1509,-0.0163,34.2000,0
190644,128921.0000,-1.3551,-2.3916,0.6313,-1.7140,0.7093,-1.9056,-1.1340,-0.1551,-1.7977,1.2695,-1.3221,-1.3478,-0.0540,-0.3521,-0.2474,-0.8185,0.6018,0.1805,0.3359,-0.2393,-0.0247,0.2682,0.5823,-0.0545,-0.5513,-0.1762,0.0122,0.4372,10.0000,0
223505,143411.0000,-2.5279,-1.1480,0.7096,-0.8862,1.9872,-1.9859,0.7914,-0.3488,-0.0026,-0.7635,-1.0237,0.1839,0.1642,-0.0804,-0.8594,0.4100,-0.8967,-0.8590,-0.8636,-0.9304,-0.4636,-0.5674,1.0827,-0.0280,0.3717,0.0780,0.0683,0.0058,0.9900,0
210793,138120.0000,1.7936,0.3190,-0.9579,3.4197,1.2466,1.8437,-0.1550,0.4301,-1.1018,1.5035,0.4828,0.5691,0.1993,0.3666,-0.7628,0.2053,-0.4166,-0.8727,-2.1844,-0.3355,0.3829,1.2946,-0.0182,-1.6112,0.0627,0.3548,0.0123,-0.0783,10.6200,0
95578,65377.0000,-0.5711,1.0900,1.4159,0.4829,0.1257,-0.1360,0.4288,0.0736,-0.8036,-0.1101,1.9375,0.6519,-0.0056,-0.1224,0.9595,-0.1864,0.4551,-0.0199,0.7905,0.0239,-0.0505,-0.2239,0.0987,0.1883,-0.7537,0.1411,-0.1297,0.1407,1.7900,0
1489,1161.0000,-1.0013,1.2682,1.9528,0.7975,-0.5745,-0.2134,0.3242,0.4881,-1.2291,-0.5420,1.4634,1.2151,0.8385,0.5362,0.5535,0.2202,-0.3027,0.4496,0.7164,0.0958,0.0079,-0.2153,-0.0106,0.5088,0.1369,-0.5214,-0.0433,0.0245,49.9900,0
256858,157895.0000,-2.9515,-8.7357,-3.1914,2.2896,-3.1903,0.9422,2.6483,-0.4563,1.3741,-1.3714,0.0220,1.0098,-0.5729,0.4793,-1.0771,0.2430,-0.2337,0.5184,0.1795,4.8974,1.1675,-2.1366,-1.8954,-0.3161,-1.3593,-0.5781,-0.4691,0.3871,2547.4700,0
2965,2508.0000,-0.9529,2.8819,-2.7903,1.3915,0.1463,-1.8491,0.2437,0.5803,0.2425,0.3028,2.3921,0.0254,-0.8922,-3.1096,0.4678,1.0613,2.8769,1.9950,-0.3519,0.5534,-0.2055,-0.1761,0.2418,0.1483,-0.2596,-0.4294,0.3651,-0.0496,0.8900,0


In [16]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='str')

In [17]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

Credit card fraud is one of the most significant financial crimes worldwide, costing billions of dollars annually. This project builds an end-to-end machine learning pipeline to **automatically detect fraudulent credit card transactions** in real time.

The dataset contains transactions made by European cardholders over **two days in September 2013**:

| Property | Detail |
|---|---|
| Total Transactions | 284,807 |
| Fraudulent Transactions | 492 (~0.17%) |
| Features | 30 (V1–V28 via PCA, + Time, Amount) |
| Target | `Class` — 0 = Legitimate, 1 = Fraud |

All features except `Time` and `Amount` have been **PCA-transformed** to protect user confidentiality.

## 2. Problem Statement

Financial institutions process **millions of transactions per day**. Manually reviewing each transaction for fraud is infeasible. The key challenges are:

- **Extreme class imbalance** — Fraudulent transactions represent only **~0.17%** of all transactions. A naive model predicting "all legitimate" achieves 99.83% accuracy but is completely useless.
- **Evolving fraud patterns** — Fraudsters constantly change tactics, making static rule-based systems ineffective.
- **Cost asymmetry** — The cost of missing a fraud (false negative) far exceeds the cost of a false alarm (false positive).
- **Speed requirement** — Fraud decisions must be made in **milliseconds** at transaction time.

> **Core Question:** *Can we build a model that reliably identifies the rare fraudulent transactions while minimising inconvenience to legitimate customers?*

## 3. Business Objective

The primary business objectives are:

1. **Maximise fraud detection (Recall)** — Catch as many fraudulent transactions as possible to minimise financial losses.
2. **Minimise false positives (Precision)** — Avoid blocking legitimate transactions, which damages customer trust and experience.
3. **Balance the trade-off** — Use metrics like **F1-score**, **PR-AUC**, and **ROC-AUC** that account for imbalance.
4. **Provide explainability** — Model decisions must be interpretable for compliance and audit purposes.

### Priority Ordering
```
Recall > Precision > Latency > Interpretability
```
A missed fraud costs the bank real money. A blocked legitimate transaction costs customer goodwill.

## 4. Stakeholders

| Stakeholder | Role | Key Concern |
|---|---|---|
| **Fraud Operations Team** | Primary users of the model | High recall — don't miss frauds |
| **Customer Experience Team** | Monitors false positives | Low false positive rate |
| **Risk & Compliance** | Regulatory oversight | Model explainability & auditability |
| **Engineering Team** | Deploys the model | Latency, scalability, reliability |
| **Finance / Management** | Business decision makers | ROI, cost savings, revenue protection |
| **End Customers** | Cardholders | Seamless experience, account security |

## 5. Business Impact

### Financial Impact
- The global cost of credit card fraud exceeded **$33 billion in 2022** (Nilson Report).
- Each undetected fraud transaction results in direct financial loss to the issuing bank.
- A **1% improvement in recall** on a bank processing 10M transactions/month with avg fraud value of $150 ≈ **$150,000+ in prevented losses**.

### Reputational Impact
- Customers who experience fraud and feel it was not caught quickly **switch banks**.
- Excessive false positives (wrongly blocked cards) also drive churn.

### Operational Impact
- Automating detection **reduces manual review workload** by fraud analysts.
- Enables real-time blocking of fraudulent transactions before settlement.

## 6. Machine Learning Formulation

### Task Type
**Binary Classification** — predict whether a transaction is fraudulent (`Class = 1`) or legitimate (`Class = 0`).

### Input Features
- `V1` – `V28`: PCA-transformed anonymised features
- `Time`: Seconds elapsed since the first transaction in the dataset
- `Amount`: Transaction amount in euros

### Target
- `Class`: `0` = Legitimate, `1` = Fraud

### Key Modelling Decisions
| Decision | Choice | Rationale |
|---|---|---|
| Imbalance handling | SMOTE, class weights, undersampling | Fraud is only 0.17% of data |
| Primary metric | **PR-AUC / F1** | ROC-AUC is misleading on imbalanced data |
| Model family | Ensemble (XGBoost, LightGBM, RF) | High performance on tabular data |
| Threshold | Tuned post-training | Business cost trade-off |
| Explainability | SHAP values | Required for compliance |

## 7. Success Criteria

The model will be considered successful if it meets the following thresholds on the **held-out test set**:

| Metric | Minimum Target | Stretch Goal |
|---|---|---|
| **Recall (Fraud)** | ≥ 0.80 | ≥ 0.90 |
| **Precision (Fraud)** | ≥ 0.70 | ≥ 0.85 |
| **F1-Score (Fraud)** | ≥ 0.75 | ≥ 0.87 |
| **PR-AUC** | ≥ 0.80 | ≥ 0.90 |
| **ROC-AUC** | ≥ 0.95 | ≥ 0.98 |

> **Note:** Recall is weighted more heavily than Precision. Missing a fraud costs more than a false alarm.

## 8. Assumptions & Limitations

### Assumptions
- The PCA-transformed features (`V1`–`V28`) retain sufficient information for fraud detection even without interpretability.
- The two-day snapshot is representative of broader transaction patterns.
- The dataset labels are accurate and were verified by domain experts.

### Limitations
| Limitation | Impact |
|---|---|
| **Anonymised features** | Cannot perform feature engineering with domain knowledge |
| **Static dataset** | Real-world fraud patterns drift over time — model will need retraining |
| **No merchant / geography data** | Important fraud signals are missing |
| **Two-day window** | Seasonal patterns and long-term behaviour not captured |
| **Single currency (EUR)** | May not generalise to other markets |

## 9. Notebook Summary

This notebook establishes the **business context** for the fraud detection project. Key takeaways:

- ✅ Fraud detection is a **high-stakes binary classification** problem
- ✅ The dataset is **severely imbalanced** (~0.17% fraud) — standard accuracy is meaningless
- ✅ **Recall is the priority metric** — missing fraud is more costly than a false alarm
- ✅ SHAP-based **explainability** is required for compliance
- ✅ Success is defined by **PR-AUC ≥ 0.80** and **Recall ≥ 0.80**

### Next Steps
Proceed to **`02_Data_Understanding.ipynb`** for:
- Dataset shape, types, and missing values
- Target class distribution analysis
- Feature statistics and correlation analysis

---
## Setup & Data Loading

In [8]:
# Load dataset — raw string to avoid backslash escape issues
DATA_PATH = Path(r"D:\Fraud Detection\data\raw\creditcard.csv")

df = pd.read_csv(DATA_PATH)
print(f"Dataset loaded: {df.shape[0]:,} rows x {df.shape[1]} columns")

Dataset loaded: 284,807 rows x 31 columns


In [9]:
# Quick class distribution check
class_counts = df['Class'].value_counts()
fraud_rate = df['Class'].mean() * 100

print("Class Distribution:")
print(f"  Legitimate (0): {class_counts[0]:,}  ({100 - fraud_rate:.2f}%)")
print(f"  Fraudulent (1): {class_counts[1]:,}  ({fraud_rate:.4f}%)")
print(f"\nImbalance ratio: {class_counts[0] / class_counts[1]:.0f}:1")

df.head()

Libraries loaded successfully.
